process/process_linux.go

```diff
@@ -322,79 +472,40 @@ func (p *Process) IsRunning() (bool, error) {
 // MemoryMaps get memory maps from /proc/(pid)/smaps
 func (p *Process) MemoryMaps(grouped bool) (*[]MemoryMapsStat, error) {
 	pid := p.Pid
-	var ret []MemoryMapsStat
+	var memMaps []MemoryMapsStat
 	smapsPath := common.HostProc(strconv.Itoa(int(pid)), "smaps")
-	contents, err := ioutil.ReadFile(smapsPath)
+
+	smaps, err := os.Open(smapsPath)
 	if err != nil {
 		return nil, err
 	}
-	lines := strings.Split(string(contents), "\n")
+	defer smaps.Close()
 
-	// function of parsing a block
-	getBlock := func(first_line []string, block []string) (MemoryMapsStat, error) {
-		m := MemoryMapsStat{}
-		m.Path = first_line[len(first_line)-1]
+	scanner := bufio.NewScanner(bufio.NewReader(smaps))
 
-		for _, line := range block {
-			if strings.Contains(line, "VmFlags") {
-				continue
+	firstLine := true
+	mms := MemoryMapsStat{}
+	for scanner.Scan() {
+		line := scanner.Text()
+		cols := strings.Fields(line)
+		if !strings.HasSuffix(cols[0], ":") {
+			if firstLine {
+				firstLine = false
+			} else {
+				memMaps = append(memMaps, mms)
 			}
-			field := strings.Split(line, ":")
-			if len(field) < 2 {
-				continue
+			mms = MemoryMapsStat{}
+			if err := mms.parseHeader(cols); err != nil {
+				return nil, err
 			}
-			v := strings.Trim(field[1], " kB") // remove last "kB"
-			t, err := strconv.ParseUint(v, 10, 64)
-			if err != nil {
-				return m, err
-			}
-
-			switch field[0] {
-			case "Size":
-				m.Size = t
-			case "Rss":
-				m.Rss = t
-			case "Pss":
-				m.Pss = t
-			case "Shared_Clean":
-				m.SharedClean = t
-			case "Shared_Dirty":
-				m.SharedDirty = t
-			case "Private_Clean":
-				m.PrivateClean = t
-			case "Private_Dirty":
-				m.PrivateDirty = t
-			case "Referenced":
-				m.Referenced = t
-			case "Anonymous":
-				m.Anonymous = t
-			case "Swap":
-				m.Swap = t
-			}
-		}
-		return m, nil
-	}
-
-	blocks := make([]string, 16)
-	for _, line := range lines {
-		field := strings.Split(line, " ")
-		if strings.HasSuffix(field[0], ":") == false {
-			// new block section
-			if len(blocks) > 0 {
-				g, err := getBlock(field, blocks)
-				if err != nil {
-					return &ret, err
-				}
-				ret = append(ret, g)
-			}
-			// starts new block
-			blocks = make([]string, 16)
 		} else {
-			blocks = append(blocks, line)
+			if err := mms.updateAttribute(cols); err != nil {
+				return nil, err
+			}
 		}
 	}
-
-	return &ret, nil
+	memMaps = append(memMaps, mms)
```

Comment:
If there are no lines, won't this append an empty stat?

Reply:
I did add a check and return an error if it's empty